### ISSUE TITLE
Questions about the code to calculate 'SupConLoss_admargin'

### ISSUE THREAD

**Issue by JixiangChen-Jimmy** [XMed-Lab/AdaCon#4]
Dear authors,
Thanks for the interesting and valuable work! During reading the code, I am confused about some code in the function 'SupConLoss_admargin' to calculate the loss. Specifically, in line 558, `anchor_dot_contrast = scale_s* (anchor_dot_contrast + adj_abdiff)` calculates the cosine similarities with adaptive margins, but what does the following code mean:
`logits_max, _ = torch.max(anchor_dot_contrast, dim=1, keepdim=True)
logits = anchor_dot_contrast - logits_max.detach()`
I did not find this operation in the formulas. Did I miss something? Can you help to provide some explanation?

**Comment by ackbar03** [comment]
Hi Jimmy, 

Thanks for your interest in our work!

For line 558, we subtract everything by  `logits_max.detach()` to help stabilize training. This does not change the actual loss since it is part of log division and the result is numerically equivalent. It is also done for example in the SupCon loss function implementation here: https://github.com/HobbitLong/SupContrast/blob/331aab5921c1def3395918c05b214320bef54815/losses.py#L73

Hope that answers your question!

Regards,
Weihang

**Comment by JixiangChen-Jimmy** [comment]
Thanks for your timely reply! It is helpful and I understand now.
